### PR TITLE
Add cw1155 specification

### LINKED ISSUE #246: No equivalent of ERC1155 standard
- ERC1155 treats fungible and non-fungible tokens equally, non-fungible tokens just have one max supply.
- Metadata and token enumerations can be handled off-chain, by indexing attributes in block results.
- Approval is set to some operator for all tokens at once, no fine-granularity control over allowance.
  Implementation might assume approval for self automatically.
- Mint/burn/transfer are mixed in one set of messages, otherwise, there'll be too many message types.

### PR DESCRIPTION
**Pull request by yihuang**
retry of https://github.com/CosmWasm/cosmwasm-plus/pull/245
fixes #246 

Design decisions:
- Fungible tokens and non-fungible tokens are treated equally, non-fungible tokens just have one max supply.
- Transfer permission is approved or disapproved to some operator over entire set of tokens. (Scoped control is defined in [ERC1761](https://eips.ethereum.org/EIPS/eip-1761), we can port that separately)
- Metadata query and token enumeration should be done by indexing attributes in block-results off-chain.
- Mint and burn are mixed with transfer/send messages to keep number of message types small, otherwise we'll need much more message types like `Mint`/`MintToContract`/`BatchMint`/`BatchMintToContract`, etc.
  In transfer/send messges, `from`/`to` are optional, a `None` `from` means minting, a `None` `to` means burning, they must not both be `None` at the same time.

### CODE REVIEW

File: packages/cw0/src/event.rs
Comment (ethanfrey) [inline packages/cw0/src/event.rs:3]: Can you add rustdoc on this?

I am sure I will see how it is used a few files down, but if someone comes to it for the first time, nice to explain.

File: packages/cw1155/Cargo.toml
Comment (ethanfrey) [inline packages/cw1155/Cargo.toml:18]: Probably should be -beta1 as well for consistency. (Not that the schema package changed)

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:12]: 1. token enumeration should be done by indexing events off-chain

This is because it is impossible in Ethereum. We can do better. How about using enumeration like in cw721 on tokens or tokens/owner? or cw20 on allowances?

Comment (ethanfrey) [reply packages/cw1155/README.md:12]: 2. Off-chain metadata via events is okay. Adding one URL there or such would be a simple addition that seems like it is used various places.

Also, the "indexing events off-chain" should reference an example and some such indexer (I think you guys have built one, so maybe a nice place to link to it and say you could build on this)

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:16]: This seems dangerous to me, as the input is JSON, and extra fields are ignored. So changing `from` to `form` turns a transfer to a mint (likely to fail anyway), but using `too` vs `to` will make you burn your tokens, rather than have the tx rejected by the chain.

@willclarktech and @maurolacy I think we discussed the issue of nullable fields changing functionality to be dangerous. Will especially in the context of CosmJS, some case when we said omitting a field from "ChangeOwner" would make it "ClearOwner" (making no one owner) could be an expensive mistake with a typo.

Can you please comment on this design here (or better look at the [corresponding type definitions](https://github.com/CosmWasm/cosmwasm-plus/pull/247/files#diff-07f199dd55ccb4e6b043c50dded9da214c860437339f9a73e0bea9ed89a85d72R12-R29)) and say if you approve or suggest ways to improve.

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:24]: The fact that you combine `Send` and `SendFrom` is nice (just provide your own address as from to make it a normal send). Just not sure to throw in Mint/MintFrom/Burn/BurnFrom here

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:51]: Very cool to specify the emitted events. We should do that on all contracts for the key events.

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:55]: Hmm.. I was going to suggest different events for mint and burn, but on reflection agree with you here. If you are searching for transfer events but the mint events are missing, you will get the wrong balances. 

This is a clever design to enable better indexing.

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:59]: Nice to emit this event. I would also store the current value of the URL in state. It is not that large. (You can max size it to 200 bytes or so to keep people from base64 encoding images into it)

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:63]: I think metadata is defined above, correct? So you can remove that part.

For enumerable, I would suggest just implementing it making use of the store iterators in the sdk. It is rather efficient.

File: packages/cw1155/src/event.rs
Comment (ethanfrey) [inline packages/cw1155/src/event.rs:13]: Very nice idea. I like it.

So, the standard defines some Event types and the contracts can just instantiate these types and "serialize" them into the response. After this PR is merged, I will look to see how to adapt other contracts to this pattern.

File: packages/cw1155/src/event.rs
Comment (ethanfrey) [inline packages/cw1155/src/event.rs:14]: Response object now has an [`add_attribute` method](https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/results/response.rs#L112-L117). Maybe this could make use of it?

```rust
fn write_attributes(&self, resp: &mut Response) {
  resp.add_attributute("action", "transfer");
  resp.add_attributute("token_id", self.token_id);
  // ...
}
```

Not that it does anything different, but seems a bit cleaner to read and sets a nice pattern for other devs. (And I would like your feedback on the method name `add_attribute`, we are still in beta, when it can be changed)

File: packages/cw1155/src/msg.rs
Comment (ethanfrey) [inline packages/cw1155/src/msg.rs:21]: -> if `env.sender` is the owner or has sufficient pre-approval.

File: packages/cw1155/src/query.rs
Comment (ethanfrey) [inline packages/cw1155/src/query.rs:21]: Again from cw721, how about iterating (page-wise) over all approvals. Here is a template for one way (and you can look at the implementation in cw721-base): https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw721/src/query.rs#L17-L25

File: packages/cw1155/src/msg.rs
Comment (ethanfrey) [inline packages/cw1155/src/msg.rs:47]: In cw721 we took a few liberties to "improve" the Ethereum design, even with the "all or nothing" approval. How about adding an [optional expiration date](https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw721/src/msg.rs#L38)?

Adding approvals on token by token basis would be nice as well, even if it is full control of the balance on that token_id. Are you concerned about performance? Or prefer the API this way? Especially iteration is MUCH more efficient in CosmWasm/Cosmos SDK than Ethereum and it is nice to embrace it.

File: packages/cw1155/src/receiver.rs
Comment (ethanfrey) [inline packages/cw1155/src/receiver.rs:13]: Since this is an interface other devs will have to support, good to be explicit.

In this case `operator` was the one who actually executed the message and `from` is the account the funds were sent from, correct? If so, please just add some RustDoc strings here.

Comment (ethanfrey) [reply packages/cw1155/src/receiver.rs:13]: Otherwise, nice job in this file.

**Review by ethanfrey — COMMENTED** [review]
Very nice and clean implementation and I really like some of the ideas. Especially events, but also merging eg. Send/SendFrom into one message type (avoids lots of duplication).

Most of the comments are minor suggestions to polish off the spec. The only real issue is combining send/burn/mint into one message. I would like a 2nd (or 3rd) opinion here. This is really about API design not coding.

Also, what do you think in general of extending the Ethereum spec to match our framework a bit I mean, you use Uint128 rather than Uint256 (128 bit is plenty large enough, and they just have 256 cuz everything is 256bit in Ethereum). It is a chance to take these same ideas and enhance them with some of the capabilities of CosmWasm (especially cheap iteration over parts of a set)

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: - Added the `Metadata` extension query apis, similar to CW721.
- Changed event name from `metadata` to `token_info`, to keep it consistent with CW721 naming.

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: I added metadata and enumerable extension similar to the ones in CW721.

File: packages/cw1155/src/event.rs
Comment (yihuang) [reply packages/cw1155/src/event.rs:14]: I'm not very good at picking english names, I think `add_attribute` is good, I changed the method name in `Event` trait to `add_attribututes` too.
Another possible choice here is `fn iter_attributes(&self, resp: &mut Response) -> impl Iterator<Item=Attribute>`.

File: packages/cw1155/src/msg.rs
Comment (yihuang) [reply packages/cw1155/src/msg.rs:21]: done

File: packages/cw1155/src/query.rs
Comment (yihuang) [reply packages/cw1155/src/query.rs:22]: Looks good, copied those from CW721 spec.

File: packages/cw1155/src/receiver.rs
Comment (yihuang) [reply packages/cw1155/src/receiver.rs:13]: done

File: packages/cw1155/src/msg.rs
Comment (yihuang) [reply packages/cw1155/src/msg.rs]: I think the only trade-off is use cases vs simplicity and on-chain resource usage.
For example, for the [example cw721 contract](https://github.com/CosmWasm/cosmwasm-plus/blob/master/contracts/cw721-base/src/state.rs#L13) to support approval by token, it needs to store a `Vec<Approval>` in `TokenInfo`, and need to do `O(n)` search when checking permission.

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: I added metadata and enumerable extensions like cw721, it's indeed more convenient for client I think.

File: packages/cw1155/Cargo.toml
Comment (yihuang) [reply packages/cw1155/Cargo.toml]: done

File: packages/cw0/src/event.rs
Comment (yihuang) [reply packages/cw0/src/event.rs:4]: done

**Comment by yihuang** [comment]
> Very nice and clean implementation and I really like some of the ideas. Especially events, but also merging eg. Send/SendFrom into one message type (avoids lots of duplication).
> 
> Most of the comments are minor suggestions to polish off the spec. The only real issue is combining send/burn/mint into one message. I would like a 2nd (or 3rd) opinion here. This is really about API design not coding.
> 
> Also, what do you think in general of extending the Ethereum spec to match our framework a bit I mean, you use Uint128 rather than Uint256 (128 bit is plenty large enough, and they just have 256 cuz everything is 256bit in Ethereum). It is a chance to take these same ideas and enhance them with some of the capabilities of CosmWasm (especially cheap iteration over parts of a set)

- Add metadata and enumeable query extension similar to cw721
- Modified ApprovalForAll similar to cw721
- Add some comments

Open issues:

- Should we mix mint/burn messages with transfer/send
- Should we support approval by token, like cw721

File: packages/cw1155/src/msg.rs
Comment (ethanfrey) [reply packages/cw1155/src/msg.rs]: Fair enough. This does add more runtime complexity. We can leave it as is.

**Comment by ethanfrey** [comment]
Thanks for the updates, looking very nice.

> Should we support approval by token, like cw721?

I don't know if this use case will be used much or at all and it does indeed add runtime overhead to every check. I am fine to leave it off.

> Should we mix mint/burn messages with transfer/send? 

This is my only outstanding concern, and I would like the comment of another developer with CosmWasm experience here. Otherwise, it looks good.

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:27]: either be (missing space)

File: packages/cw1155/README.md
Comment (ethanfrey) [inline packages/cw1155/README.md:46]: Actually interesting point - start_after does not include that value itself. It is great for pagination, but not good for finding one in particular. Maybe you do want the other query as well to just get one.

**Review by ethanfrey — COMMENTED** [review]
Looks good minus the discussion on separating Transfer/Mint/Burn.

If you wish to reduce types, another thought would be to combine TransferFrom and SendFrom.

eg.

```rust
/// This handles contract call or user account
SendFrom {
  from: HumanAddr, // either env.sender self or another account sender is approved on
  to: HumanAddr, // an external account or contract 
  token_id: TokenId,
  value: Uint128,
  // if msg is None, just transfer the tokens.
  // if msg is Some(m), then call a Cw1155 receiver on a contract with the `to` address (after moving the token)
  msg: Option<Binary>,
}
```

And then add other types for minting and burning (or at least burning).
This would allow you to reduce the number of types and address my concern about possibly mixing burning and send.

If you want to still keep `from: None` to mean `mint` I have less issues than with burn. In the end maybe you can end up with the followng types:

* `SendFrom` - mints or transfers from my account or transfers from other account. If there is a message, it will assume the target is a contract and invoke the cw1155 receiver on it.
* `BatchSendFrom` - like SendFrom but with a list of tokens
* `BurnFrom` - burns one token either from my account or an account i was authorized on
* `ApproveAll`
* `RevokeAll`

File: packages/cw1155/README.md
Comment (willclarktech) [reply packages/cw1155/README.md]: What about minting/burning if the null address is provided, like Ethereum's `0x0000000000000000000000000000000000000000`? Then at least you have to specify it. Otherwise I'd suggest `mint`/`burn` boolean options to be explicit.

**Comment by yihuang** [comment]
> ```
>   // if msg is None, just transfer the tokens.
>  // if msg is Some(m), then call a Cw1155 receiver on a contract with the `to` address (after moving the token)
>  msg: Option<Binary>,
> ```

I think it needs to be `Option<Option<Binary>>`:
- `None`, normal recipient address
- `Some(None)`, recipient address is contract, but send `None` as extra msg.
- `Some(Some(msg))`, recipient address is contract, send extra msg.

> `BurnFrom` - burns one token either from my account or an account i was authorized on

I guess we probably also want `BatchBurnFrom`?

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: added `ApprovedForAllItem`.

**Comment by ethanfrey** [comment]
> I think it needs to be Option<Option<Binary>>:

`Option<Option<Binary>>` and `Option<Binary>` have the same JSON representation.

There is no real use case when we send a message to a contract with data `None`. You can always send `Some(b"{}")` for the cases when it doesn't matter (and ensure you have a valid JSON object as body)

File: packages/cw1155/src/query.rs
Comment (ethanfrey) [inline packages/cw1155/src/query.rs:32]: I would rename to `IsApprovedForAll` that would suggest a boolean as a response.

The other one is a bit vague (calling it `ListApprovedForAll` would be more explicit), but I think clear enough.

**Comment by yihuang** [comment]
> > I think it needs to be Option<Option>:
> 
> `Option<Option<Binary>>` and `Option<Binary>` have the same JSON representation.
> 
> There is no real use case when we send a message to a contract with data `None`. You can always send `Some(b"{}")` for the cases when it doesn't matter (and ensure you have a valid JSON object as body)

But the receiver msg contains `Option<Binary>`, do you think we should remove the `Option` here too?

File: packages/cw1155/src/query.rs
Comment (yihuang) [reply packages/cw1155/src/query.rs:32]: > I would rename to `IsApprovedForAll` that would suggest a boolean as a response.

that's better, thanks.

> The other one is a bit vague (calling it `ListApprovedForAll` would be more explicit), but I think clear enough.

I think current overall naming conversion is using nouns for message branch names, so I guess we can hold this one for now?

**Comment by ethanfrey** [comment]
> But the receiver msg contains Option<Binary>, do you think we should remove the Option here too?

You are right. I designed the cw20 receiver interface before implementing any contracts. It would be good to remove Option there.

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: done

File: packages/cw1155/README.md
Comment (yihuang) [reply packages/cw1155/README.md]: Now, mint/burn messages are separated.

**Comment by yihuang** [comment]
- Merged TransferFrom into SendFrom
- Separated out Mint/Burn messages

**Review by ethanfrey — APPROVED** [review]
Thank you for the changes

File: packages/cw1155/README.md
Comment (ThienLK1) [inline packages/cw1155/README.md]: In event.rs i only see MetadataEvent, not found token_info.
And i dont see any api using this event.
I just want add uri for token, but dont find any api do this?

Comment (yihuang) [reply packages/cw1155/README.md]: metadata modification is not part of the standard API, different contracts can implement their own APIs to do that I think.

Comment (ThienLK1) [reply packages/cw1155/README.md]: Do you mean base on ERC1155Metadata?
Do you think it can fulfill what's missing?

**Pull request merged**
